frontends/common/constantFolding.h

```diff
@@ -23,36 +23,97 @@ limitations under the License.
 #include "frontends/p4/typeChecking/typeChecker.h"
 
 namespace P4 {
-// Can also be run before type checking, so it only uses types if
-// they are available.
+
+/** Implements a pass that statically evaluates many constant
+ * expressions.
+ *
+ * Currently the following expressions are treated as constants:
+ *  - `IR::Constant`
+ *  - `IR::BoolLiteral`
+ *  - `IR::ListExpression` where each element is also a constant
+ *  - `IR::TypeNameExpression`
+ *
+ * The pass can be invoked either with or without the reference and
+ * type maps. When this information is not available, then several
+ * cases revert to the identify function.
+ *
+ * The pass also eliminates arbitrary-precision integer literals,
+ * replacing them with literals that have explicit widths.
+ *
+ * @pre: None
+ *
+ * @post: Ensures that
+ *    - some constant expressions are eliminated (TODO: specify)
+ *    - arbitrary-precision integers replaced with fixed-width integers (TODO: confirm)
+ */
 class DoConstantFolding : public Transform {
  protected:
-    const ReferenceMap* refMap;  // if null no 'const' values can be resolved
-    TypeMap* typeMap;  // if null we have no types; updated for new constants
+    /// Used to resolve IR nodes to declarations.
+    /// If `null`, then `const` values cannot be resolved
+    const ReferenceMap* refMap;
+
+    /// Used to resolve nodes to their types.
+    /// If `null`, then type information is not available
+    TypeMap* typeMap;
+
+    /// Set to `true` iff type information is available
     bool typesKnown;
-    bool warnings;  // if true emit warnings
-    // maps expressions and declarations to their constant values
+
+    /// If `true` then emit warnings
+    bool warnings;
+
+    /// Maps IR nodes to constants
     std::map<const IR::Node*, const IR::Expression*> constants;
 
  protected:
-    // returns nullptr if the expression is not a constant
+    /** Check whether @p expr is a constant. @returns @p expr if a
+     *  constant and `nullptr` if not.  The result is guaranteed to have one
+     *  of the following types:
+     *  - `IR::Constant`
+     *  - `IR::BoolLiteral`
+     *  - `IR::ListExpression`, where each element is also a constant
+     *  - `IR::TypeNameExpression`
```

Comment:
I don't think this is correct. This can be a Member(TypeName), where TypeName denotes an enum type.